ast/src/type.rs

```diff
@@ -1,7 +1,76 @@
 use dbg_pls::DebugPls;
+use std::fmt::{Debug, Display, Formatter, Write};
+use std::ops::Deref;
 
 #[derive(Debug, Clone, PartialEq, DebugPls)]
-pub struct Type<'a> {
+pub enum Type<'a> {
+    ///A Simple type with optional parameters (`A`, `A[V]`)
+    Simple(SimpleType<'a>),
+    ///A Lambda declaration with of form `(A, B, ...) => Out`
+    Lambda(LambdaType<'a>),
+    ///Either `()` or `Unit`, representing a void type
+    Unit,
+}
+
+#[derive(Debug, Clone, PartialEq, DebugPls)]
+pub struct SimpleType<'a> {
     pub name: &'a str,
     pub params: Vec<Type<'a>>,
 }
+
+#[derive(Debug, Clone, PartialEq, DebugPls)]
+pub struct LambdaType<'a> {
+    pub inputs: Vec<Type<'a>>,
+    pub output: Box<Type<'a>>,
+}
+
+impl<'a> Display for Type<'a> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Type::Simple(m) => Display::fmt(m, f),
+            Type::Lambda(p) => Display::fmt(p, f),
+            Type::Unit => write!(f, "Unit"),
+        }
+    }
+}
+
+///helper function to write a type list format in a given formatter
+fn display_type_list<'a>(
```

Comment:
This function could definitely be moved to a separate file as the code is also shared by https://github.com/moshell-lang/moshell/blob/d91964823796d7255daded79f9767c408d4de5ac/parser/src/aspects/type.rs#L99